src/amuse/ext/orbital_elements.py

```diff
@@ -27,81 +28,228 @@ def newton(f,x0,fprime=None,args=(),tol=1.48e-8,maxiter=50):
+    try:
+        unit = angle.unit
+    except:
+        angle = angle | default_unit
+    return angle
 
-    cos_long_asc_nodes = numpy.cos(longitude_of_the_ascending_node)
-    sin_long_asc_nodes = numpy.sin(longitude_of_the_ascending_node)
 
-    ### alpha is a unit vector directed along the line of node ###
-    alphax = cos_long_asc_nodes*cos_arg_per - sin_long_asc_nodes*sin_arg_per*cos_inclination
-    alphay = sin_long_asc_nodes*cos_arg_per + cos_long_asc_nodes*sin_arg_per*cos_inclination
+def equal_length_array_or_scalar(
+        array, length=1, mode="continue"
+        ):
+    """
+    Returns 'array' if its length is equal to 'length'.
+    If this is not the case, returns 'array' anyway if it is a scalar,
+    or the first value of the array if its length is different from 'length'.
+    If mode is "warn", issues a warning if this happens; if mode is "exception"
+    raises an exception in this case.
+    """
+    try:
+        array_length = len(array)
+        if array_length == length:
+            return array
+        else:
+            if mode == "warn":
+                warnings.warn("Length of array is not equal to %i. Using only\
+                        the first value." % length)
+            elif mode == "exception":
+                raise Exception("Length of array is not equal to %i. This is\
+                not supported." % length)
+            return array[0]
+    except:
+        if mode == "warn":
+            warnings.warn("Not an array, continuing with scalar.")
+        elif mode == "exception":
+            raise Exception("Not an array, this is not supported.")
+        return array
+
+
+def center_of_mass_array(
+        vectors,
+        primary_mass,
+        secondary_mass,
+        ):
+    """
+    Returns array of center_of_mass vectors, where primaries are considered to
+    be at (0,0,0) and secondaries at 'vectors'.
+    """
+    total_mass = (primary_mass + secondary_mass).reshape(
+            (len(primary_mass), 1)
+            )
+    center_of_mass_array = (
+            (
+                vectors
+                * secondary_mass.reshape(
+                    (len(secondary_mass), 1)
+                    )
+                )
+            / total_mass
+            )
+    return center_of_mass_array
+
+
+def generate_binaries_from_orbital_elements(
+        primary_mass,
+        secondary_mass,
+        semi_major_axis,
+        eccentricity=0 | units.rad,
+        true_anomaly=0 | units.rad,
+        inclination=0 | units.rad,
+        longitude_of_the_ascending_node=0 | units.rad,
+        argument_of_periapsis=0 | units.rad,
+        G=nbody_system.G
+        ):
+    """
+    returns two particlesets, which contain the primaries and the secondaries
+    in binary pairs.
+    """
+    number_of_primaries = len(primary_mass)
+    number_of_secondaries = len(secondary_mass)
+
+    # mass arrays need to be the same length
+    if number_of_secondaries != number_of_primaries:
```

Comment:
maybe add another acceptable case: number_of_primaries==1 (or input not a vector) and then generate one set assuming independent (zeroish mass) orbiters